File: 9012/9012.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    9012.py                                            :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/02/25 17:06:47 by jihoolee          #+#    #+#              #
#    Updated: 2024/02/25 17:13:20 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

def is_VPS(PS: str) -> bool:
    s: list[str] = []

    for c in PS:
        if c == '(':
            s.append('(')
        else:
            try:
                s.pop()
            except Exception:
                return False
    return True if len(s) == 0 else False


def main() -> None:
    for _ in range(int(input())):
        ps: str = input()
        print('YES' if is_VPS(ps) else 'NO')


if __name__ == '__main__':
    main()

File: 9012/9012.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   9012.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/02/25 16:41:24 by jihoolee          #+#    #+#             */
/*   Updated: 2024/02/25 17:09:59 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <string>
#include <stack>

bool is_VPS(const std::string& PS) {
  std::stack<char>  s;

  for (size_t idx = 0; idx < PS.length(); ++idx) {
    if (PS[idx] == '(') {
      s.push('(');
    } else {
      if (s.empty())
        return false;
      s.pop();
    }
  }
  if (!s.empty())
    return false;
  return true;
}

int main(void) {
  int         case_cnt;
  std::string PS;
  std::string result;

  std::cin >> case_cnt;
  for (int i = 0; i < case_cnt; ++i) {
    std::cin >> PS;
    result = is_VPS(PS) ? "YES": "NO";
    std::cout << result << std::endl;
  }
  return 0;
}
